torch/csrc/api/include/torch/detail/TensorDataContainer.h

```diff
@@ -0,0 +1,272 @@
+// 2. `at::ArrayRef` of supported tensor data types.
+// 3. `std::vector` of supported tensor data types.
+//
+// At any time, a `TensorDataContainer` object represents one of the following:
+//
+// 1. A scalar with value `scalar()` and type `scalar_type()`.
+// 2. A Tensor represented in `std::initializer_list<TensorDataContainer>` form,
+//    with value `init_list()`, Tensor scalar type `scalar_type()`, and Tensor sizes `sizes()`.
+// 3. A Tensor represented in `at::Tensor` form, with value `tensor()`, scalar type `scalar_type()`,
+//    and Tensor sizes `sizes()`.
+//
+// All the infrastructure here is mostly to support converting an arbitrarily nested braced-init-list
+// to the equivalent Tensor successfully. Consider the following example:
+//
+// `torch::tensor({{1}, {2}})`
+//
+// Here is the code path that it goes through:
+//
+// `at::Tensor tensor(detail::TensorDataContainer init_list_tensor)`
+//
+// which calls:
+//
+// `TensorDataContainer({{1}, {2}})`
+//
+// which matches to the `TensorDataContainer(std::initializer_list<TensorDataContainer>)` constructor,
+// and in an attempt to convert `{1}` and `{2}` to `TensorDataContainer`, it calls the following:
+//
+// `TensorDataContainer({1})`  (same call path happens for `{2}`, and we'll just focus on `{1}` here)
+//
+// At this point, theoretically there are two plausible ways for `{1}` to be matched to one of the
+// constructors of `TensorDataContainer`:
+//
+// 1. It can be a list-initialization of a scalar value, thus matching `TensorDataContainer(int value)`.
+// 2. It can be converted to `std::initializer_list<TensorDataContainer>`, thus matching
+//    `TensorDataContainer(std::initializer_list<TensorDataContainer>)`.
+//
+// How does the compiler decide which one to choose? According to `https://en.cppreference.com/w/cpp/language/list_initialization`,
+// braced-init-list always prefers the constructor that takes `std::initializer_list`. Hence we happily
+// move forward with constructor #2, and it calls the following:
+//
+// `TensorDataContainer(1)`
+//
+// Now it matches `TensorDataContainer(int value)`, which stores `1` as a scalar value. All is good.
+//
+// Note that `torch::tensor({{1}, {2}})` can also match another previously existing function overload:
+// `torch::tensor(at::ArrayRef<int> values)`, because `{1}` and `{2}` can be treated as
+// a list-initialization of an `int` value. However, this will produce a Tensor with sizes `{2}`,
+// but we actually want a Tensor with sizes `{2, 1}`. In order to avoid matching this function overload,
+// we removed the function overload and moved the ability to convert `at::ArrayRef<T>` (and similarly `std::vector<T>`)
+// into `TensorDataContainer`, and since for braced-init-list the `TensorDataContainer(std::initializer_list<TensorDataContainer>)`
+// constructor is always preferred over all other constructors, it will take the `std::initializer_list` path
```

Comment:
nit: It's kinda hard to read lines on github when they're over 80-100 lines long (this one is 132), but don't worry about changing it

Reply:
Fixed